packages/liveblocks-node/src/client.ts

```diff
@@ -69,36 +121,84 @@ export class Liveblocks {
   }
 
   /** @internal */
-  private async post(
-    path: `/${string}`,
-    json: Record<string, unknown>
-  ): Promise<Response> {
+  private async post(path: URLSafeString, json: Json): Promise<Response> {
     const url = urljoin(this._baseUrl, path);
     const headers = {
       Authorization: `Bearer ${this._secret}`,
       "Content-Type": "application/json",
     };
 
     const fetch = await fetchPolyfill();
-    return fetch(url, {
+    const res = await fetch(url, {
       method: "POST",
       headers,
       body: JSON.stringify(json),
     });
+    if (!res.ok) {
+      const text = await res.text();
+      throw new HttpError(res.status, text);
+    }
+    return res;
```

Comment:
With these generic `.post()`, `.get()`, `.put()`, etc helpers, you're either throwing the response if it's a non-200 response, or _returning_ the response. However, this isn't obvious from the call site perspective if you look at the return type, because it just says `Response`. So from the call site it's unclear that those responses will only ever be 2xx responses, and `res.ok` will never be `true` on the returned value.

I'd suggest to therefore return the result of `res.json()` here and change the return type for these helpers to `Promise<Json>` instead, to help make that intent and behavior more clear just by looking at the types.

Reply:
I like this and makes sense! I do have some concerns though but not major ones: this assumes that all responses will be in JSON format and if the server sends a response in a different format, this will fail or will need additional handling. I noticed inside `identifyUser`, we make a `POST` request and then parse the response body as text before returning it. 

https://github.com/liveblocks/liveblocks/blob/ea2b50212f51c08eba046da122ed4c45b01438c4/packages/liveblocks-node/src/client.ts#L272-L284

This was one of the reasons why I initially opted out of checking `res.status` inside the generic `POST`, `GET` methods so that public methods calling these internal methods had more flexibility in terms of what they did with the response but there was repetition involved with checking for `res.status` for each method, which also wasn't desirable. 

Another minor concern is that returning only JSON means that we lose access to other details of the response, such as headers, status code, or status text, which is completely fine with the current setup, but something to keep in mind.

What are your thoughts @nvie ?

Reply:
Yeah, it's a tough trade-off to make.

All _success_ responses from our API should (and will always) be in JSON format I think? If CF has a disruption, they may return a HTML-based 503 page, in which case there is no JSON payload, but that case is handled by your `res.ok` check already. I think all of our success responses should just always return JSON, to make handling this easy.

Although some of our endpoints may return an empty response body (e.g. in the case of an [HTTP 204](https://github.com/liveblocks/liveblocks-cloudflare/blob/2ee8995fc5646e71bf7af5a8c3a2a17412568f64/packages/liveblocks-cloudflare/src/api/v2/get-rooms.ts#L77-L86)), then JSON parsing would still fail. Maybe we should simply use our `tryParseJson` helper instead (from `@liveblocks/core`), and use this instead:

```tsx
return tryParseJson(await res.text()) ?? null;
```

I don't think we actually need any response info (from headers or status fields). All information this client should return to the user is coming from the response payload. If this changes in the future, we could expose a "raw" version that would retain the rich response info and maybe let the error handling be dealt with by the user, but for now I'd say YAGNI.

Reply:
That is true. All our success responses do return JSON and the `tryParseJson` should gracefully deal with cases where we send an empty response body. With this change, how do you suggest we update the `identifyUser`. Do we do something like this instead?

Before:
```ts
try {
  // this.post() used to return Promise<Response>
  const resp = await this.post(path, {
    userId,
    groupIds,
  });

  return {
    status: normalizeStatusCode(resp.status),
    body: await resp.text(),
  };
}
```

After:
```ts
try {
  // this.post() now returns Promise<Json | null>
  const json = await this.post(path, {
    userId,
    groupIds,
  });

  return {
    status: 200,
    body: JSON.stringify(json),
  };
}
```

Reply:
Sure, that works! I wouldn't worry about the `JSON.stringify(JSON.parse(...))` that's happening there effectively perf-wise — if that's what you mean. You can stringify/parse a typical response many 100000's times in one second, it's nothing compared to the async/networking overhead here.

I think this code is really readable this way 👍

Reply:
Hmmm. I looked more into this and I think this might not work for all cases. For e.g. when requesting for (binary) Yjs document, the server sends binary data. See [link](https://github.com/liveblocks/liveblocks-cloudflare/blob/2ee8995fc5646e71bf7af5a8c3a2a17412568f64/packages/liveblocks-cloudflare/src/room.ts#L608-L617).

```ts
const update = Y.encodeStateAsUpdate(doc);
return new Response(update, {
  headers: { "Content-Type": "application/octet-stream" },
  status: 200,
});
```

This needs to be read as `res.arrayBuffer()` from the client. This is making me more inclined towards keeping the generic `post`, `get` internal methods only do the fetching and letting the methods that consume this method deal with the rest of the logic as they see fit despite the repetition that is involved (e.g. handling non-2xx responses) until we come up with an approach that fits our situation well. Also, these are all implementation details, so we can always go back and fine tune these later.

So, something like:

```ts
private async post(path: URLSafeString, json: Json): Promise<Response> {
  const url = urljoin(this._baseUrl, path);
  const headers = {
    Authorization: `Bearer ${this._secret}`,
    "Content-Type": "application/json",
  };

  const fetch = await fetchPolyfill();
  return await fetch(url, {
    method: "POST",
    headers,
    body: JSON.stringify(json),
  });
}

public async getRoom(roomId: string): Promise<RoomInfo> {
  const res = await this.get(url`/v2/rooms/${roomId}`);

  if (!res.ok) {
    const text = await res.text();
    throw new HttpError(res.status, text);
  }

  return (await res.json()) as Promise<RoomInfo>;
}
```